synapse/lib/agenda.py

```diff
@@ -0,0 +1,614 @@
+            if self.incunit is None:
+                largest_req = min(self.reqdict.keys())
+                tmpunit = _NextUnitMap[largest_req]
+                if tmpunit is None:  # required a year and we're already there
+                    return 0.0
+                # Unless we're going to the next day of week, increment by 1 unit of the next larger unit
+                tmpincval = self.reqdict.get(TimeUnit.DAYOFWEEK, 1)
+            else:
+                tmpunit = self.incunit
+                tmpincval = self.incval
+            newdt = self._inc(tmpunit, tmpincval, self.reqdict, lastdt, newdt)
+            assert newdt > lastdt
+        return newdt.timestamp()
+
+    def _inc(self, incunit, incval, reqdict, origdt, dt):
+        '''
+        Return a datetime incremented by incunit * incval
+        '''
+        if incunit == TimeUnit.YEAR:
+            return dt.replace(year=dt.year + incval)
+        if incunit == TimeUnit.MONTH:
+            newyear = dt.year
+            absmonth = dt.month + incval - 1
+            newmonth = absmonth % 12 + 1
+            newyear += absmonth // 12
+            daysinmonth = calendar.monthrange(newyear, newmonth)[1]
+            dayofmonthreq = reqdict.get(TimeUnit.DAYOFMONTH)
+            if dayofmonthreq is not None:
+                newday = _dayofmonth(dayofmonthreq, newmonth, newyear)
+            else:
+                newday = min(daysinmonth, dt.day)
+            return dt.replace(day=newday, month=newmonth, year=newyear)
+        if incunit == TimeUnit.DAY:
+            return dt + datetime.timedelta(days=incval)
+        if incunit == TimeUnit.DAYOFWEEK:
+            # incval in this case means next day of week whose weekday matches incval (0-6)
+            days = (6 + incval - dt.weekday()) % 7 + 1
+            newdt = dt + datetime.timedelta(days=days)
+            assert newdt.weekday() == incval
+            return newdt
+        if incunit == TimeUnit.HOUR:
+            return dt + datetime.timedelta(hours=incval)
+        if incunit == TimeUnit.MINUTE:
+            return dt + datetime.timedelta(minutes=incval)
+        else:
+            assert 0, 'Invalid incunit'  # pragma: no cover
+
+class _Appt:
+    '''
+    A single entry in the Agenda:  a storm query to run in the future, potentially more than once
+
+    Each such entry has a list of ApptRecs.  Each time the appointment is scheduled, the nexttime of the appointment is
+    the lowest nexttime of all its ApptRecs.
+    '''
+    def __init__(self, iden, recur, indx, query, username, recs, nexttime=None):
+        self.iden = iden
+        self.recur = recur # does this appointment repeat
+        self.indx = indx  # incremented for each appt added ever.  Used for nexttime tiebreaking for stable ordering
+        self.query = query  # query to run
+        self.username = username  # user to run query as
+        self.recs = recs  # List[ApptRec]
```

Comment:
Comment: the individual combos outstanding. These get deleted in non-recurring appointments.